src/mtev_cluster.c

```diff
@@ -223,54 +255,118 @@ mtev_cluster_check_timeouts() {
   packed_time |= (my_boot_time.tv_usec & 0xffffff); /* 3 bytes */
   packed_time = htonll(packed_time);
   MEMWRITE(&packed_time, sizeof(packed_time));
-  seq = htonll(cluster->config_seq);
-  MEMWRITE(&seq, sizeof(seq));
+  my_seq = htonll(cluster->config_seq);
+  MEMWRITE(&my_seq, sizeof(my_seq));
+  MEMWRITE(&number_of_payloads, 1);
+
+  assert(HEART_BEAT_HDR_LEN == MEMWRITE_WRITTEN);
+
+  if(number_of_payloads != 0) {
+    // write header (pointer table)
+    while(mtev_hash_next(&cluster->hb_payloads, &payload_iter, &key, &klen, &vdata)) {
+      hb_payload = vdata;
+      MEMWRITE(&hb_payload->app_id, 1);
+      MEMWRITE(&hb_payload->key, 1);
+      MEMWRITE(&hb_payload->data_len, 1);
+    }
+    // write payload
+    while(mtev_hash_next(&cluster->hb_payloads, &payload_iter2, &key, &klen, &vdata)) {
+      hb_payload = vdata;
+      MEMWRITE(hb_payload->data, hb_payload->data_len);
+    }
+  }
 
   /* TODO: Support registered composers */
-  MEMWRITE_RETURN;
+  return MEMWRITE_WRITTEN;
+}
+static mtev_boolean
+mtev_cluster_store_payload(mtev_cluster_node_t *node, const void* payload, uint16_t payload_length, uint8_t number_of_payloads) {
+  assert(payload != NULL);
+  node->number_of_payloads = number_of_payloads;
+  if(node->payload && node->payload_length < payload_length) {
+    node->payload = realloc(node->payload, payload_length);
+  } else if(node->payload == NULL) {
+    node->payload = malloc(payload_length);
+  }
+  if(node->payload == NULL || payload == NULL) {
+    return mtev_false;
+  }
+  node->payload_length = payload_length;
+  memcpy(node->payload, payload, payload_length);
+
+  return mtev_true;
 }
 static int
-mtev_cluster_info_process(void *payload, int len, void *c) {
+mtev_cluster_info_process(void *msg, int len, void *c) {
+  uint8_t version;
+  uint8_t understood_version;
   mtev_cluster_node_t *sender;
-  MEMREAD_DECL(payload, len);
+  MEMREAD_DECL(msg, len);
   mtev_cluster_t *cluster = c;
   uuid_t nodeid;
   struct timeval boot_time;
   u_int64_t packed_time;
   u_int64_t seq;
-  /* payloads from other cluster members (including me) arrive here */
+  void* payload = NULL;
+  uint8_t number_of_payloads;
+  uint16_t payload_len;
+  mtev_boolean payload_changed = mtev_true;
+
+  /* messages from other cluster members (including me) arrive here */
+  MEMREAD(&version, 1);
+  understood_version = version & 0x0F;
+  version = version >> 4;
   MEMREAD(nodeid, UUID_SIZE);
   MEMREAD(&packed_time, sizeof(packed_time));
   packed_time = ntohll(packed_time);
   boot_time.tv_sec = ((packed_time >> 24) & 0x000000ffffffffffULL);
   boot_time.tv_usec = (packed_time & 0xffffff);
   MEMREAD(&seq, sizeof(seq));
   seq = ntohll(seq);
+  MEMREAD(&number_of_payloads, 1);
+
+  payload_len = len-MEMREAD_BYTES_READ;
+  MEMGET(payload, payload_len);
+
+  // We currently support only one version
+  assert (version == understood_version && version == HEARTBEAT_MESSAGE_VERSION);
```

Comment:
Another Node could crash us. If they aren't the same we should just ignore the packet (and log)